Sources/NIOExtras/LengthFieldBasedFrameDecoder.swift

```diff
@@ -0,0 +1,130 @@
+///
+///     +-----+------+
+///     | BCD | FGHI |
+///     +-----+------+
+///
+/// 'A' and 'E' will be the headers and will not be passed forward.
+///
+
+public final class LengthFieldBasedFrameDecoder: ByteToMessageDecoder {
+
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    
+    public var cumulationBuffer: ByteBuffer?
+    
+    private let lengthFieldLength: ByteLength
+    private let lengthFieldEndianness: Endianness
+    
+    /// Create `LengthFieldBasedFrameDecoder` with a given frame length.
+    ///
+    /// - parameters:
+    ///    - lengthFieldLength: The length of the field specifying the remaining length of the frame.
+    ///    - lengthFieldEndianness: The endianness of the field specifying the remaining length of the frame.
+    ///
+    public init(lengthFieldLength: ByteLength, lengthFieldEndianness: Endianness = .big) {
+        self.lengthFieldLength = lengthFieldLength
+        self.lengthFieldEndianness = lengthFieldEndianness
+    }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        
+        guard let lengthFieldSlice = buffer.readSlice(length: self.lengthFieldLength.length) else {
+            return .needMoreData
+        }
+
+        // convert the length field to an int specifying the length
+        guard let lengthFieldValue = self.frameLength(for: lengthFieldSlice) else {
+            return .needMoreData
+        }
+
+        guard let contentsFieldSlice = buffer.readSlice(length: lengthFieldValue) else {
+            return .needMoreData
+        }
+
+        ctx.fireChannelRead(self.wrapInboundOut(contentsFieldSlice))
+
+        return .continue
+    }
+    
+    public func handlerRemoved(ctx: ChannelHandlerContext) {
+        if let buffer = cumulationBuffer, buffer.readableBytes > 0 {
+            ctx.fireErrorCaught(NIOExtrasErrors.LeftOverBytesError(leftOverBytes: buffer))
+        }
+    }
+
+    ///
+    /// Decodes the specified region of the buffer into an unadjusted frame length. The default implementation is
+    /// capable of decoding the specified region into an unsigned 8/16/32/64 bit integer.
+    /// - parameters:
+    ///    - buffer: The buffer containing the integer frame length
+    ///
+    private func frameLength(for buffer: ByteBuffer) -> Int? {
+
+        switch self.lengthFieldLength {
+        case .one:
+            return buffer.getInteger(at: 0, endianness: self.lengthFieldEndianness, as: UInt8.self).map { Int($0) }
+        case .two:
+            return buffer.getInteger(at: 0, endianness: self.lengthFieldEndianness, as: UInt16.self).map { Int($0) }
+        case .four:
+            return buffer.getInteger(at: 0, endianness: self.lengthFieldEndianness, as: UInt32.self).map { Int($0) }
+        case .eight:
+            return buffer.getInteger(at: 0, endianness: self.lengthFieldEndianness, as: UInt64.self).map { Int($0) }
```

Comment:
This will fail on 32 bit platforms. You’ll need to use `UInt64` to parse the length information.

Reply:
@weissi It's a bit more awkward than that. This value is being passed to `readSlice`, so it's possible that on 32-bit platforms we'll be asked to read more from the slice than is possible to represent in the 32-bit `Int`. More importantly, `readSlice` doesn't take a `UInt64`, it takes an `Int`, so whatever is returned here has to be converted into an `Int` anyway.

Frankly, I think the edge-case isn't worth worrying about: if a 32-bit machine is asked to read more than `Int.max` from a frame something bad is about to happen anyway.

Reply:
What about if we guard against the use of 64bit length field on a 32bit platform in the constructor? I can add a comment next to it and it saves anyone getting into a pickle but gets this through without a major refactor. I am happy to do the work if required, it just sounds like it is slightly out of the scope of this PR.

Reply:
@tigerpixel / @Lukasa I think we should guard in the constructor, otherwise this is a good way to mount a DOS attack. The attacker can just send `0xffffffff` as the length to a 32 bit system and frameLength will always return `nil` so it'll always return `.needMoreData` which will cause the `ByteToMessageDecoder` to read arbitrary amounts of data into memory.

Reply:
Added.